src/Command/Phar/SelfUpdateCommand.php

```diff
@@ -0,0 +1,196 @@
+    {
+        return $this->console->getVersion();
+    }
+
+    /**
+     * @throws \RuntimeException
+     */
+    private function getMostRecentAvailableVersion(): string
+    {
+        return trim($this->download('version.txt'));
+    }
+
+    /**
+     * @throws \RuntimeException
+     */
+    private function fetchPharContents(): string
+    {
+        $expectedSha384 = trim($this->download('concrete.sig'));
+        $phar = $this->download('concrete.phar');
+        if ($this->output->isVerbose()) {
+            $this->output->write('Checking signature of the PHAR file... ');
+        }
+        $actualSha384 = hash('sha384', $phar);
+        if ($expectedSha384 !== $actualSha384) {
+            throw new RuntimeException(
+<<<EOT
+The downloaded phar file is corrupt!
+Expected SHA-384: {$expectedSha384}
+Actual SHA-384  : {$actualSha384}
+EOT
+            );
+        }
+        if ($this->output->isVerbose()) {
+            $this->output->outputDone("passed.\n");
+        }
+
+        return $phar;
+    }
+
+    /**
+     * @throws \RuntimeException
+     */
+    private function download(string $file): string
+    {
+        if ($this->output->isVerbose()) {
+            $this->output->write("Downloading {$file}... ");
+        }
+        $url = sprintf(self::LASTESTRELEASE_PATTERN, $file);
+        $context = stream_context_create([
+            'http' => [
+                'follow_location' => 1,
+                'ignore_errors' => false,
+            ],
+        ]);
+        $error = 'Unknown error';
+        set_error_handler(static function(int $errno, string $errstr) use (&$error): void {
+            $error = trim($errstr);
+            if ($error === '') {
+                $error = "Unknown error (code: {$errno})";
+            }
+        }, -1);
+        $result = file_get_contents($url, false, $context);
+        restore_error_handler();
+        if ($result === false) {
+            throw new RuntimeException("Failed to download {$url}:\n{$error}");
+        }
+        if ($this->output->isVerbose()) {
+            $length = strlen($result);
+            $this->output->outputDone("done ({$length} bytes).\n");
+        }
+
+        return $result;
+    }
+
+    /**
+     * @throws \RuntimeException
+     */
+    private function updateMyPhar(string &$pharContents): void
+    {
+        if ($this->output->isVerbose()) {
+            $this->output->write("Updating PHAR... ");
+        }
+        $pharPath = $this->getPharPath();
+        set_error_handler(static function(): void {}, -1);
+        $pharPermissions = fileperms($pharPath);
+        restore_error_handler();
+        $error = 'Unknown error';
+        set_error_handler(static function(int $errno, string $errstr) use (&$error): void {
+            $error = trim($errstr);
+            if ($error === '') {
+                $error = "Unknown error (code: {$errno})";
+            }
+        }, -1);
+        $savedBytes = file_put_contents($pharPath, $pharContents);
```

Comment:
Would it be better to save this new phar in a temp file then copy it over top the old one? It might have less chance of leaving you in a broken state if the write gets interrupted for any reason

Reply:
Should we rely on `sys_get_temp_dir()`?

Reply:
Thinking a bit more: it's better to save in the same dir as the phar: f12fd856a4f0ff803dc39e8ccbc0253c0b6fe312